diff --git a/hipamd/include/hip/nvidia_detail/nvidia_hip_runtime_api.h b/hipamd/include/hip/nvidia_detail/nvidia_hip_runtime_api.h
index 0522cc8e..8678403b 100644
--- a/hipamd/include/hip/nvidia_detail/nvidia_hip_runtime_api.h
+++ b/hipamd/include/hip/nvidia_detail/nvidia_hip_runtime_api.h
@@ -39,6 +39,7 @@ THE SOFTWARE.
 #define CUDA_11030 11030
 #define CUDA_11040 11040
 #define CUDA_11060 11060
+#define CUDA_12000 12000
 
 #ifdef __cplusplus
 extern "C" {
@@ -2668,17 +2669,21 @@ inline static hipError_t hipCtxGetFlags(unsigned int* flags) {
     return hipCUResultTohipError(cuCtxGetFlags(flags));
 }
 
+#if CUDA_VERSION < CUDA_12000
 inline static hipError_t hipCtxDetach(hipCtx_t ctx) {
     return hipCUResultTohipError(cuCtxDetach(ctx));
 }
+#endif  // CUDA_VERSION < CUDA_12000
 
 inline static hipError_t hipDeviceGet(hipDevice_t* device, int ordinal) {
     return hipCUResultTohipError(cuDeviceGet(device, ordinal));
 }
 
+#if CUDA_VERSION < CUDA_12000
 inline static hipError_t hipDeviceComputeCapability(int* major, int* minor, hipDevice_t device) {
     return hipCUResultTohipError(cuDeviceComputeCapability(major, minor, device));
 }
+#endif  // CUDA_VERSION < CUDA12000
 
 inline static hipError_t hipDeviceGetName(char* name, int len, hipDevice_t device) {
     return hipCUResultTohipError(cuDeviceGetName(name, len, device));
@@ -2742,9 +2747,11 @@ inline static hipError_t hipModuleGetFunction(hipFunction_t* function, hipModule
     return hipCUResultTohipError(cuModuleGetFunction(function, module, kname));
 }
 
+#if CUDA_VERSION < CUDA_12000
 inline static hipError_t hipModuleGetTexRef(hipTexRef* pTexRef, hipModule_t hmod, const char* name){
     return hipCUResultTohipError(cuModuleGetTexRef(pTexRef, hmod, name));
 }
+#endif  // CUDA_VERSION < CUDA_12000
 
 inline static hipError_t hipFuncGetAttributes(hipFuncAttributes* attr, const void* func) {
     return hipCUDAErrorTohipError(cudaFuncGetAttributes(attr, func));
@@ -2792,6 +2799,7 @@ inline static hipError_t hipFuncSetCacheConfig(const void* func, hipFuncCache_t
     return hipCUDAErrorTohipError(cudaFuncSetCacheConfig(func, cacheConfig));
 }
 
+#if CUDA_VERSION < 12000
 __HIP_DEPRECATED inline static hipError_t hipBindTexture(size_t* offset,
                                                          struct textureReference* tex,
                                                          const void* devPtr,
@@ -2805,6 +2813,7 @@ __HIP_DEPRECATED inline static hipError_t hipBindTexture2D(
     const hipChannelFormatDesc* desc, size_t width, size_t height, size_t pitch) {
     return hipCUDAErrorTohipError(cudaBindTexture2D(offset, tex, devPtr, desc, width, height, pitch));
 }
+#endif  // CUDA_VERSION <= 12000
 
 inline static hipChannelFormatDesc hipCreateChannelDesc(int x, int y, int z, int w,
                                                         hipChannelFormatKind f) {
@@ -2837,10 +2846,12 @@ inline static hipError_t hipGetTextureObjectResourceDesc(hipResourceDesc* pResDe
     return hipCUDAErrorTohipError(cudaGetTextureObjectResourceDesc( pResDesc, textureObject));
 }
 
+#if CUDA_VERSION < CUDA_12000
 __HIP_DEPRECATED inline static hipError_t hipGetTextureAlignmentOffset(
     size_t* offset, const struct textureReference* texref) {
     return hipCUDAErrorTohipError(cudaGetTextureAlignmentOffset(offset,texref));
 }
+#endif  // CUDA_VERSION < CUDA_12000
 
 inline static hipError_t hipGetChannelDesc(hipChannelFormatDesc* desc, hipArray_const_t array)
 {
@@ -2864,6 +2875,7 @@ inline static hipError_t hipModuleLaunchCooperativeKernel(hipFunction_t f, unsig
                                                            sharedMemBytes, stream,kernelParams));
 }
 
+#if CUDA_VERSION < CUDA_12000
 inline static hipError_t hipLaunchCooperativeKernelMultiDevice(hipLaunchParams* launchParamsList,
                                                  int  numDevices, unsigned int  flags) {
     return hipCUDAErrorTohipError(cudaLaunchCooperativeKernelMultiDevice(launchParamsList, numDevices, flags));
@@ -2876,6 +2888,7 @@ inline static hipError_t hipModuleLaunchCooperativeKernelMultiDevice(
     return hipCUResultTohipError(cuLaunchCooperativeKernelMultiDevice(launchParamsList,
                                                                       numDevices, flags));
 }
+#endif
 
 inline static hipError_t hipImportExternalSemaphore(hipExternalSemaphore_t* extSem_out,
                                       const hipExternalSemaphoreHandleDesc* semHandleDesc) {
@@ -3086,6 +3099,7 @@ inline static hipError_t hipOccupancyMaxActiveBlocksPerMultiprocessorWithFlags(
                                                                  blockSize, dynamicSMemSize, flags));
 }
 
+#if CUDA_VERSION < CUDA_12000
 template <class T, int dim, enum cudaTextureReadMode readMode>
 inline static hipError_t hipBindTexture(size_t* offset, const struct texture<T, dim, readMode>& tex,
                                         const void* devPtr, size_t size = UINT_MAX) {
@@ -3128,6 +3142,7 @@ __HIP_DEPRECATED inline static hipError_t hipBindTextureToArray(
     struct texture<T, dim, readMode>& tex, hipArray_const_t array) {
     return hipCUDAErrorTohipError(cudaBindTextureToArray(tex, array));
 }
+#endif  // CUDA_VERSION < CUDA_12000
 
 template <class T>
 inline static hipChannelFormatDesc hipCreateChannelDesc() {
@@ -3164,6 +3179,7 @@ inline static hipError_t hipTexObjectGetTextureDesc(HIP_TEXTURE_DESC* pTexDesc,
     return hipCUResultTohipError(cuTexObjectGetTextureDesc(pTexDesc, (CUtexObject)texObject));
 }
 
+#if CUDA_VERSION < CUDA_12000
 __HIP_DEPRECATED inline static hipError_t hipTexRefSetAddressMode(hipTexRef hTexRef, int dim, hipAddress_mode am){
     return hipCUResultTohipError(cuTexRefSetAddressMode(hTexRef,dim,am));
 }
@@ -3191,6 +3207,7 @@ __HIP_DEPRECATED inline static hipError_t hipTexRefSetFlags(hipTexRef hTexRef, u
 __HIP_DEPRECATED inline static hipError_t hipTexRefSetArray(hipTexRef hTexRef, hiparray hArray, unsigned int Flags){
     return hipCUResultTohipError(cuTexRefSetArray(hTexRef,hArray,Flags));
 }
+#endif  // CUDA_VERSION < 12000
 
 inline static hipError_t hipArrayCreate(hiparray* pHandle, const HIP_ARRAY_DESCRIPTOR* pAllocateArray){
     return hipCUResultTohipError(cuArrayCreate(pHandle, pAllocateArray));
