diff --git a/include/hip/amd_detail/device_library_decls.h b/include/hip/amd_detail/device_library_decls.h
index 02228705..55817c8a 100644
--- a/include/hip/amd_detail/device_library_decls.h
+++ b/include/hip/amd_detail/device_library_decls.h
@@ -118,10 +118,10 @@ extern "C" __device__ uint64_t __ockl_fprintf_append_string_n(uint64_t msg_desc,
                                                               uint64_t length, uint32_t is_last);

 // Introduce local address space
-#define __local __attribute__((address_space(3)))
+#define __device_local __attribute__((address_space(3)))

 #ifdef __HIP_DEVICE_COMPILE__
-__device__ inline static __local void* __to_local(unsigned x) { return (__local void*)x; }
+__device__ inline static __device_local void* __to_local(unsigned x) { return (__device_local void*)x; }
 #endif //__HIP_DEVICE_COMPILE__
 
 // Using hip.amdgcn.bc - sync threads
