diff --git a/include/hip/nvidia_detail/nvidia_hip_runtime_api.h b/include/hip/nvidia_detail/nvidia_hip_runtime_api.h
index cb9b23aa..bbda413b 100644
--- a/include/hip/nvidia_detail/nvidia_hip_runtime_api.h
+++ b/include/hip/nvidia_detail/nvidia_hip_runtime_api.h
@@ -2136,7 +2136,7 @@ inline static hipError_t hipModuleGetFunction(hipFunction_t* function, hipModule
 }

 inline static hipError_t hipModuleGetTexRef(hipTexRef* pTexRef, hipModule_t hmod, const char* name){
-    hipCUResultTohipError(cuModuleGetTexRef(pTexRef, hmod, name));
+    return hipCUResultTohipError(cuModuleGetTexRef(pTexRef, hmod, name));
 }

 inline static hipError_t hipFuncGetAttributes(hipFuncAttributes* attr, const void* func) {
