diff --git a/utils/bazel/llvm-project-overlay/clang/BUILD.bazel b/utils/bazel/llvm-project-overlay/clang/BUILD.bazel
index 2156e55bc..28f592e5c 100644
--- a/utils/bazel/llvm-project-overlay/clang/BUILD.bazel
+++ b/utils/bazel/llvm-project-overlay/clang/BUILD.bazel
@@ -709,8 +709,9 @@ cc_library(
         # least bad approach. Using `includes` is *specifically* problematic for
         # this library because it contains files that collide easily with system
         # headers such as `CXXABI.h`.
-        "-I$(GENDIR)/external/llvm-project/clang/lib/AST",
-        "-I$(GENDIR)/external/llvm-project/clang/lib/AST/Interp",
+        "-I$(GENDIR)/external/rules_ll.llvm_project_overlay.llvm-project/clang/lib/AST",
+        "-I$(GENDIR)/external/rules_ll.llvm_project_overlay.llvm-project/clang/lib/AST/Interp",
+
     ],
     textual_hdrs = [
         "include/clang/AST/AttrImpl.inc",
@@ -1464,7 +1465,8 @@ genrule(
     outs = [hdr.replace("lib/Headers/", "staging/include/") for hdr in builtin_headers],
     cmd = """
        for src in $(SRCS); do
-         relsrc=$${src/*external\\/llvm-project\\/clang\\/lib\\/Headers\\/}
+         relsrc=$${src/*external\\/rules_ll*.llvm_project_overlay.llvm-project\\/clang\\/lib\\/Headers\\/}
+
          target=$(@D)/staging/include/$$relsrc
          mkdir -p $$(dirname $$target)
          cp $$src $$target
