diff --git a/clang-tools-extra/clang-tidy/tool/run-clang-tidy.py b/clang-tools-extra/clang-tidy/tool/run-clang-tidy.py
index ec1fd6679..afe2bfd28 100755
--- a/clang-tools-extra/clang-tidy/tool/run-clang-tidy.py
+++ b/clang-tools-extra/clang-tidy/tool/run-clang-tidy.py
@@ -90,7 +90,7 @@ def make_absolute(f, directory):
 
 def get_tidy_invocation(f, clang_tidy_binary, checks, tmpdir, build_path,
                         header_filter, allow_enabling_alpha_checkers,
-                        extra_arg, extra_arg_before, quiet, config_path,
+                        extra_arg, extra_arg_before, quiet, config_file_path,
                         config, line_filter, use_color):
   """Gets a command line for clang-tidy."""
   start = [clang_tidy_binary]
@@ -121,8 +121,8 @@ def get_tidy_invocation(f, clang_tidy_binary, checks, tmpdir, build_path,
   start.append('-p=' + build_path)
   if quiet:
       start.append('-quiet')
-  if config_path:
-      start.append('--config-file=' + config_path)
+  if config_file_path:
+      start.append('--config-file=' + config_file_path)
   elif config:
       start.append('-config=' + config)
   start.append(f)
@@ -194,7 +194,7 @@ def run_tidy(args, clang_tidy_binary, tmpdir, build_path, queue, lock,
                                      tmpdir, build_path, args.header_filter,
                                      args.allow_enabling_alpha_checkers,
                                      args.extra_arg, args.extra_arg_before,
-                                     args.quiet, args.config_path, args.config,
+                                     args.quiet, args.config_file, args.config,
                                      args.line_filter, args.use_color)
 
     proc = subprocess.Popen(invocation, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
@@ -304,7 +304,7 @@ def main():
                                      None, build_path, args.header_filter,
                                      args.allow_enabling_alpha_checkers,
                                      args.extra_arg, args.extra_arg_before,
-                                     args.quiet, args.config_path, args.config,
+                                     args.quiet, args.config_file, args.config,
                                      args.line_filter, args.use_color)
     invocation.append('-list-checks')
     invocation.append('-')
